File: VB/Data/DataProvider.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Web
Imports System.Web.SessionState

Namespace GridViewCustomValidationInTemplates

	Public NotInheritable Class DataProvider
		Private Const SessionKey As String = "MySampleData"

		Private Sub New()
		End Sub
		Private Shared ReadOnly Property Session() As HttpSessionState
			Get
				Return HttpContext.Current.Session
			End Get
		End Property

		Public Shared Function GetData() As IList(Of Record)
			Dim data As IList(Of Record) = TryCast(Session(SessionKey), IList(Of Record))
			If data Is Nothing Then
				data = CreateData()
				Session(SessionKey) = data
			End If
			Return data
		End Function
		Public Shared Function FindRecordByID(ByVal id As Integer) As Record
			Dim data As IList(Of Record) = GetData()
			For Each record As Record In data
				If record.ID = id Then
					Return record
				End If
			Next record
			Return Nothing
		End Function

		Private Shared Function CreateData() As IList(Of Record)
			Return New List(Of Record)(New Record() { New Record(1), New Record(2, Nothing), New Record(3, 1), New Record(4, 2, 1, 3), New Record(5, 0, 1, 3), New Record(6, 4, 1, 3), New Record(7, 4, 1, Nothing), New Record(8, 4, 1, Nothing), New Record(9, Nothing, 1, 2), New Record(10, -2, -1, 1) })
		End Function
	End Class

End Namespace

File: CS/Data/DataProvider.cs
using System;
using System.Collections.Generic;
using System.Web;
using System.Web.SessionState;

namespace GridViewCustomValidationInTemplates {

    public static class DataProvider {
        private const string SessionKey = "MySampleData";

        private static HttpSessionState Session {
            get { return HttpContext.Current.Session; }
        }

        public static IList<Record> GetData() {
            IList<Record> data = Session[SessionKey] as IList<Record>;
            if(data == null) {
                data = CreateData();
                Session[SessionKey] = data;
            }
            return data;
        }
        public static Record FindRecordByID(int id) {
            IList<Record> data = GetData();
            foreach(Record record in data)
                if(record.ID == id)
                    return record;
            return null;
        }

        private static IList<Record> CreateData() {
            return new List<Record>(new Record[] {
                /* new Record(id, value = null, min = null, max = null) */
                new Record(1),
                new Record(2, null),
                new Record(3, 1),
                new Record(4, 2, 1, 3),
                new Record(5, 0, 1, 3),
                new Record(6, 4, 1, 3),
                new Record(7, 4, 1, null),
                new Record(8, 4, 1, null),
                new Record(9, null, 1, 2),
                new Record(10, -2, -1, 1)
            });
        }
    }

}